src/middlewared/middlewared/plugins/device_/vrrp_events.py

```diff
@@ -1,10 +1,196 @@
+            VrrpObjs.fifo_thread.unpause()
+        if VrrpObjs.event_thread is not None and VrrpObjs.event_thread.is_alive():
+            VrrpObjs.event_thread.unpause()
+
+
+class VrrpEventThread(Thread):
+
+    def __init__(self, *args, **kwargs):
+        super(VrrpEventThread, self).__init__()
+        self.middleware = kwargs.get('middleware')
+        self.event_queue = kwargs.get('queue')
+        self.shutdown_event = Event()
+        self.pause_event = Event()
+        self.grace_period = 0.5
+        self.user_provided_timeout = kwargs.get('timeout') or 2
+        self.max_wait = self.user_provided_timeout + self.grace_period
+        self.settle_time = (self.max_wait / 2) + self.grace_period
+        self.max_rapid_settle_time = 5
+        self.rapid_event_settle_time = min(2 * self.user_provided_timeout, self.max_rapid_settle_time)
+
+    def shutdown(self):
+        self.shutdown_event.set()
+
+    def format_fifo_msg(self, msg):
+        if any((
+            not isinstance(msg, dict),
+            not msg.get('event'),
+            len(msg['event'].split()) != 4,
+            not msg.get('time'),
+        )):
+            LOGGER.error('Ignoring unexpected VRRP event message: %r', msg)
+            return
+
+        try:
+            info = msg['event'].split()
+            ifname = info[1].split('_')[0].strip('"')  # interface
+            event = info[2]  # the state that is being transititoned to
+        except Exception:
+            LOGGER.error('Failed parsing vrrp message', exc_info=True)
+            return
+        else:
+            if event not in ('MASTER', 'BACKUP'):
+                return
+
+        return {'ifname': ifname, 'event': event, 'time': msg['time']}
+
+    def pause(self):
+        self.pause_event.set()
+
+    def unpause(self):
+        self.pause_event.clear()
+
+    @property
+    def user_provided_timeout(self):
+        return self.__upt
+
+    @user_provided_timeout.setter
+    def user_provided_timeout(self, value):
+        self.__upt = value
+
+    def run(self):
+        set_name('vrrp_event_thread')
+        LOGGER.info('vrrp event thread started')
+        last_event, backoff = None, False
+        while not self.shutdown_event.is_set():
+            if self.pause_event.is_set():
+                # A BACKUP event has to migrate all the VIPs
+                # off of the controller and the only way to
+                # do that is to restart the vrrp service.
+                # However, restarting the VRRP service triggers
+                # more BACKUP events for the other interfaces
+                # so we will pause this thread while we become
+                # the backup controller and then unpause after
+                last_event = None
+                self.event_queue.clear()
+                sleep(0.2)
+
+            try:
+                event = self.event_queue[-1]
+                this_event = self.format_fifo_msg(event)
+            except IndexError:
```

Comment:
For future, you can probably create a generic "wakeup" Event() and block loop with `wait(timeout=self.max_wait)` to avoid having to sleep. Then have it wake up either via shutdown() or being set from the VrrpFifoThread.